### ISSUE TITLE
Problem with totality check when returning functions

### ISSUE THREAD

**Issue by ngrimm** [FStarLang/FStar#379]
Only the first of the following definitions verifies:

```
val test1 : nat -> nat -> Tot nat
let rec test1 x y = if x = 0 then y else test1 0 y

val test2 : nat -> Tot (nat -> Tot nat)
let rec test2 x y = if x = 0 then y else test2 0 y

val test3 : nat -> Tot (nat -> Dv nat)
let rec test3 x y = if x = 0 then y else test3 0 y 
```

The verifications of the second and third definition fail with

```
Subtyping check failed; expected type _6_318:nat{(_6_318 << %[x; y])}; got type int
```

, so FStar fails to show that the arguments decrease in the recursive call.

This is unexpected, as 
1. the definition of `test1` is (according to the tutorial) only shorthand for the definition of `test2`, so there shouldn't be any different behaviour
2. in `test3` the totality check is not even required

**Comment by vkanne-msft** [comment]
Not reproducible with --universes flag.

**Comment by catalin-hritcu** [comment]
This has only gotten *worse* with `--universes` since we get less informative, but still spurious, error messages:
```
**[hritcu@detained bug-reports]$ fstar.exe --universes bug379.fst
/home/hritcu/Projects/fstar/pub/examples/bug-reports/bug379.fst(7,47-7,48): failed to prove a pre-condition
/home/hritcu/Projects/fstar/pub/examples/bug-reports/bug379.fst(10,47-10,48): failed to prove a pre-condition
Error: 2 errors were reported (see above)
**
```

**Comment by nikswamy** [comment]
Curious why @vkanne-msft decided to close this ...

**Comment by vkanne-msft** [comment]
Don't remember.will check again.

Get Outlook for Android<https://aka.ms/ghei36>



On Fri, Jun 24, 2016 at 1:21 PM -0700, "nikswamy" <notifications@github.com<mailto:notifications@github.com>> wrote:


Curious why @vkanne-msft<https://github.com/vkanne-msft> decided to close this ...

-

**Comment by catalin-hritcu** [comment]
This got better again, but still not working. Now only `test2` fails with the following error message:
```
./bug379.fst(7,47-7,48): (Error) Subtyping check failed; expected type (uu___#559:nat{(precedes uu___@0 (LexCons x (LexCons y LexTop)))}); got type int (see ./bug379.fst(7,0-7,50))
```

**Comment by mtzguido** [comment]
After the fix to #1017, only the first one verifies, as expected. Although one can work around that restriction and do
```
module Dv

val test1 : nat -> nat -> Tot nat
let rec test1 x y = if x = 0 then y else test1 0 y

val test2 : nat -> Tot (nat -> Tot nat)
let rec test2 x = let f y = if x = 0 then y else test2 0 y in f

val test3 : nat -> Tot (nat -> Dv nat)
let rec test3 x = let f y = if x = 0 then y else test3 0 y in f
```
all of which verify.
